Ap/zend 304

a-petrini opened the pull request:
This pull request adds several checks when building a sidechain tx commitment tree, managing the errors returned by CCTPcrypto lib.
In particular, the validity check is performed:

- before accepting a tx/cert in the mempool
- creating a block via getblocktemplate()
- validating an incoming block

drgora commented on src/miner.cpp:
```diff
@@ -809,9 +845,15 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn,  unsigned int nBlo
         {
-            pblock->hashScTxsCommitment = pblock->BuildScTxsCommitment(view);
+            pblock->hashScTxsCommitment.SetNull();
+            pblock->BuildScTxsCommitment(view, pblock->hashScTxsCommitment);
```
We might assert that result is `true` here?

a-petrini replied:
Yes, addressed in e6db47d33

ptagl commented on qa/rpc-tests/sc_big_commitment_tree.py:
```diff
@@ -0,0 +1,254 @@
+        self.sync_all()
+
+        # Ask for a new block template
+        time.sleep(5)
```
Can we get rid of this `sleep` now that we have the `wait_for_block_template()` function?

```suggestion
```

a-petrini replied:
Addressed in e6db47d33

ptagl commented on qa/rpc-tests/sc_big_commitment_tree.py:
```diff
@@ -0,0 +1,254 @@
+        self.sync_all()
+
+        # Ask for a new block template
+        time.sleep(5)
```
Can we get rid of this sleep now that we have the wait_for_block_template() function?

```suggestion
```

a-petrini replied:
Addressed in e6db47d33

ptagl commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -2491,3 +2500,409 @@ TEST(CctpLibrary, TestCustomFieldsValidationFix)
+    ASSERT_TRUE(cmtObj2.add(tx2));
+    ASSERT_FALSE(cmtObj2.add(tx3));
+
+    // tx with CSW after tx with FT
```
Is it possible that the comment here should be reversed?

```suggestion
    // tx with FT after tx with CSW
```

a-petrini replied:
Correct. Addressed in e6db47d33

ptagl commented on src/main.cpp:
```diff
@@ -1142,6 +1142,13 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+        return state.DoS(100, error("%s(): SidechainTxsCommitmentBuilder failed\n", __func__),
+                        CValidationState::Code::INVALID, "bad-txns-txscommitmenttree-fail");
+    }
+
```
I see here that the check prevents a single transaction from exceeding the capacity of the commitment tree, and that in case the sender node is banned, it's a very important check.

I wonder if we could find another place for this, for instance at the end of `Sidechain::checkTxSemanticValidity()`, so that it is performed both for the mempool and the block validation.

JackPiri replied:
@ptagl It seems to me this is already available for both `AcceptTxToMemoryPool` and `CheckBlock`.

a-petrini replied:
Checks on the commitment tree are performed in different places, depending on which entity we are currently checking, covering all the possibilities:
- for incoming blocks, we test the correctness of the txs commitment `CheckBlock()`, by building a `SidechainTxsCommitmentBuilder` and progressively adding all the txs an certs contained within
- we do the same thing in `CreateNewBlock()` for blocks we are generating i.e. through `getblocktemplate()`
- for an incoming tx to be added to the mempool, we do this check in `AcceptTxToMemoryPool()->CheckTransaction()->CheckTransactionWithoutProofVerification()`
- for an incoming cert to be added to the mempool, we do this check directly in `AcceptCertificateToMemoryPool()`, as for this check we need a valid view to extract the sidechain from `CSidechainMap`, given its sidechain id; so although this check being theoretically pure semantic, we cannot move it to `Sidechain::checkTxSemanticValidity()`

drgora commented on src/sc/sidechainTxsCommitmentBuilder.h:
```diff
@@ -42,6 +42,28 @@ class SidechainTxsCommitmentBuilder
+        std::map<uint256, CommitmentBuilderGuardAliveCounter>  cbgaMap;
+        std::map<uint256, CommitmentBuilderGuardCeasedCounter> cbgcMap;
+        // The following values MUST be aligned with those specified in CCTPlib!
+        const int   SC_LIMIT = 4096;
```
I think it would be good to take these values directly from CCTPlib. I understand this requires updating external libraries, so it is ok to leave it for another PR.

Also, I believe it makes sense to make these constants `static` to save some memory.

a-petrini replied:
I totally agree with this one, as I am aware that manually keep all the values aligned is error-prone.
However I think that these symbols are not exposed in zendoo_CCTPlib.a, hence we need to modify the behaviour of that lib.
Agree on moving this task to another PR.

Also, e6db47d33 changed variable declarations to `static`

ptagl commented on src/main.cpp:
```diff
@@ -1317,6 +1324,15 @@ MempoolReturnValue AcceptCertificateToMemoryPool(CTxMemPool& pool, CValidationSt
+                                CValidationState::Code::INVALID, "bad-sc-cert-txscommitmenttree-fail");
+                return MempoolReturnValue::INVALID;
+            }
+
```
Similar to what I mentioned in the previous comment, would it be possible to move this check to `Sidechain::checkCertSemanticValidity()`?

a-petrini replied:
See previous answer

ptagl commented on src/rpc/mining.cpp:
```diff
@@ -731,7 +731,8 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
             CCoinsViewCache view(pcoinsTip);
-            pblock->hashScTxsCommitment = pblock->BuildScTxsCommitment(view);
+            pblock->hashScTxsCommitment.SetNull();
+            pblock->BuildScTxsCommitment(view, pblock->hashScTxsCommitment);
```
If I understand correctly, if we reach this line the commitment tree root should be computed without errors.

What about adding an `assert` on the return value just to be sure that we don't return a valid block with an invalid commitment tree in case of malfunctions?

a-petrini replied:
Agree, it could be quite useful. Addressed in e6db47d33

ptagl replied:
Thanks! Sorry for nitpicking here, but given that we may want in the future to completely get rid of all the asserts in the code, what do you think about keeping the call to `BuildScTxsCommitment()` outside of the `assert` and just checking the return code?

ptagl commented on src/sc/sidechainTxsCommitmentBuilder.cpp:
```diff
@@ -151,15 +151,49 @@ bool SidechainTxsCommitmentBuilder::add_fwt(const CTxForwardTransferOut& ccout,
+        LogPrint("sc", "%s():%d - scTxsCommitment building failed: too many sidechains, when adding scId[%s].\n",
+            __func__, __LINE__, ccout.GetScId().ToString());
+        return false;
+    }
```
Would it be possible that this check fails in the following scenario?
- A transaction with `SC_LIMIT` sidechains
- One _Forward Transfer Output_ for each sidechain
- Then we add a new _Forward Transfer Output_ for an **existing sidechain**

If that's the case, should we perform the check only if the sidechain ID is not already present in the tree?

JackPiri replied:
I'd also consider moving from equality check to inequality check (even if I understand fwt are added one by one).

Possibly a check like this:

```
if ((cbg.cbgaMap.size() + cbg.cbgcMap.size()) + (1 - cbg.cbgaMap.count(ccout.GetScId())) > cbg.SC_LIMIT)
```

could handle the scenario described above?

Similar comment applies to `add_bwtr` and `add_cert` (`cbg.cbgaMap`) and `add_csw` (`cbg.cbgcMap`).

a-petrini replied:
Yes, the `if` condition was not correct. Addressed in e6db47d33. 
@JackPiri I prefer to state it in this way:
`if ((cbg.cbgaMap.count(ccout.GetScId()) == 0) && ((cbg.cbgaMap.size() + cbg.cbgcMap.size()) == cbg.SC_LIMIT))`
which in my opinion is more clear.

ptagl replied:
Thanks, the current checks seem to be working properly now.

I consider this comment as resolved, I just wanted to propose an additional improvement: given that this check is performed at least 3 or 4 times in different parts of the commitment builder, can we encapsulate it somehow in the class (for instance having functions like `ceasedTreeContains()`, `aliveTreeContains()`, `totalSize()`, `isFull()`, etc.)?

JackPiri replied:
I consider this comment as resolved.

ptagl commented on src/sc/sidechainTxsCommitmentBuilder.cpp:
```diff
@@ -276,19 +381,54 @@ bool SidechainTxsCommitmentBuilder::add_cert(const CScCertificate& cert, const S
+
+    CommitmentBuilderGuardAliveCounter& counterRef = cbg.cbgaMap[cert.GetScId()];
+
+    // Check against the number of BWT already present in the sc subtree.
```
```suggestion
    // Check against the number of certificates already present in the sc subtree.
```

a-petrini replied:
Corrected in e6db47d33

ptagl reviewed (changes requested):
In general, the pull request seems fine to me, I've only left a few comments for possible improvements, thanks!

As we talked about offline, we still need to bind the additional checks to the fork manager for a controlled introduction.

Additionally, I was wondering if it could be useful to add a couple of more tests (or extending the existing ones):
-  a test that adds 4095 FTs, one for each sidechain, and then adds a second FT for any sidechain (this test should fail now but should pass after fixing the checks in the Commitment Builder)
- a test that tries to completely fill a commitment tree with 4095 sidechain, 4095 certificates, 4095 forward transfers, etc. If it takes too long to generate proofs, we could try to generate fake ones.

drgora commented on qa/rpc-tests/sc_big_commitment_tree.py:
```diff
@@ -0,0 +1,254 @@
+        assert(block_template is not None)
+        current_sc_txs_commitment_tree_root = block_template['scTxsCommitment']
+        # Check it has been included in the block candidate
+        assert_true(last_tx in [tx['hash'] for tx in block_template['transactions']], "The last TX should not be included in the block candidate!!!")
```
I believe this should be "The last TX should be included in the block candidate!!!"

a-petrini replied:
Yes, addressed in e6db47d33

JackPiri commented on src/primitives/block.h:
```diff
@@ -183,7 +183,7 @@ class CBlock : public CBlockHeader
 
     // return the sc txs commitment calculated as described in zendoo paper. It is based on contribution from
```
Maybe the description of this method can be adapted like:

```
// Builds the sc txs commitment calculated as described in zendoo paper. It is based on contribution from
// sidechains-related txes and certificates contained in this block. Returns the building result
```

a-petrini replied:
Reworded in e6db47d33

JackPiri commented on src/miner.cpp:
```diff
@@ -528,6 +529,15 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn,  unsigned int nBlo
+    // with the txs and certs that are progressively added to the candidate block.
+    // Each time this happens we check for failures that may be caused by:
+    // - reaching the SC subtree limit (currently max 4096 sidechains)
+    // - reaching the FT / BWT / CERT / CSW limit in each subtree (currently max 4096 in each category subtree per sc)
```
Very minor point: for `FT / BWT / CERT / CSW` the limit currently is `4095`; I wonder if it is maybe more maintainable to drop the numeric values and maybe just reference the associated constant variables names.

a-petrini replied:
Reworded in e6db47d33

drgora reviewed (commented):
I left few comments, none of which is fundamental.
This looks good for merge, once [this](https://github.com/HorizenOfficial/zen/pull/515/files#r1119076368) comment is solved. Thanks.

JackPiri commented on src/main.cpp:
```diff
@@ -1317,6 +1324,15 @@ MempoolReturnValue AcceptCertificateToMemoryPool(CTxMemPool& pool, CValidationSt
+            // checking txs commitment tree validity
+            SidechainTxsCommitmentBuilder scCommitmentBuilder;
+            if (!scCommitmentBuilder.add(cert, view)) {
+                state.Invalid(
```
I have two questions here:

1. shouldn't this piece of code be inside `CheckCertificate` rather than in `AcceptCertificateToMemoryPool`? in this way the code would be more aligned with transaction analogous (`CheckTransactionWithoutProofVerification`) and even more important the check would involve both mempool-validation as well as block-validation

2. why in this case failure results in a call to `state.Invalid` while on the transaction analogous (`CheckTransactionWithoutProofVerification`) failure results in a call to `state.DoS`?

a-petrini replied:
1. See [this answer](https://github.com/HorizenOfficial/zen/pull/515#discussion_r1121465652)
2. Yes, agree. Changed back to `DoS` in e6db47d33

JackPiri commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -25,6 +25,13 @@ extern unsigned char ReverseBitsInByte(unsigned char input);
 
 using namespace blockchain_test_utils;
 
+// Keep them aligned with those defined in CCTPlib!
```
Can we leverage `getCBG()` method to insert a static assert in order to make sure constant values in `test_libzendoo.cpp` are aligned with those in` sidechainTxsCommitmentBuilder.h`? Or at least add a check within the test execution?

JackPiri commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -2491,3 +2500,409 @@ TEST(CctpLibrary, TestCustomFieldsValidationFix)
+
+    SidechainTxsCommitmentBuilder cmtObj;
+
+    CTransaction scCreationTx = txCreationUtils::createNewSidechainTxWith(CAmount(10), /*height*/10);
```
Very low important and coming from previous commit; I think the comment `/*height*/` should be removed (the value `10` being `epochLength`); similar at lines 584 and 2824.

a-petrini replied:
Yes, corrected in e6db47d33

JackPiri commented on qa/rpc-tests/sc_big_commitment_tree.py:
```diff
@@ -0,0 +1,254 @@
+#!/usr/bin/env python3
```
I see in this file (line 248) and in `sc_big_commitment_tree_getblockmerkleroot.py` (lines 49, 51, 57) the critical values 4095/4096 being used as numeric constant rather than constant variables.
Maybe a quick change can be to make them global/shared within Python modules (like for example `ForkHeights`).
As a wider consideration, I see those values being duplicated in cpp production code, cpp test code (gtest) and python test code, in addition to rust code (cctplib); this can be a valid point supporting a change to cctplib and mccryptolib in order to avoid duplications and possible silent bugs (something related is under internal evaluation).

a-petrini replied:
Corrected in e6db47d33001

JackPiri reviewed (changes requested):
The PR is overall fine to me, but I have left some comments (some to be addressed, other with minor importance)

JackPiri reviewed (approved):
PR approved, thanks.

a-petrini committed:
fix: add checks to CreateNewBlock(..) and getblocktemplate() to avoid mismatch in scTxsCommitment and txs in a candidate block

a-petrini committed:
fix: added commitment guard check when connecting a block [ConnectBlock()]

a-petrini committed:
fix: New CommitmentBuilderGuard limits up to 4095 FT/CERT/BWTR/CSW for sidechain subtree in txsCommitmentTreeBuilder

- Added CommitmentBuilderGuard class to keep track of how many leaves have been added to a SidechainTxsCommitmentBuilder
- Added CCTPlib failing management when adding leaves
- Added BuildScTxsCommitmentGuard in CBlock structure

a-petrini committed:
test: added sc_big_commitment_tree.py and sc_big_commitment_tree_getblockmerkleroot.py

a-petrini committed:
CI: fix prepare stage in Ubuntu Bionic

a-petrini committed:
fix: improved getblockmerkleroots()

a-petrini committed:
test: added gtest suite for ScCommitmentBuilder/Guard

Also fixed a couple of failing tests that required specific information (not random)

a-petrini committed:
fix: added txs commitment tree guard for tx and cert in mempool

ptagl commented on src/miner.cpp:
```diff
@@ -809,9 +871,16 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn,  unsigned int nBlo
         {
-            pblock->hashScTxsCommitment = pblock->BuildScTxsCommitment(view);
+            pblock->hashScTxsCommitment.SetNull();
+            bool retValtxsComm = pblock->BuildScTxsCommitment(view, pblock->hashScTxsCommitment);
```
Looking at the call to `BuildScTxsCommitment()`, would it make sense to improve encapsulation by making the function immediately update the CBlock member `hashScTxsCommitment` without passing it as a parameter?

a-petrini replied:
Agree. `BuildScTxsCommitment` is now `UpdateScTxsCommitment` which directly modifies `hashScTxsCommitment` member variable

ptagl commented on src/sc/sidechainTxsCommitmentBuilder.cpp:
```diff
@@ -161,6 +161,14 @@ bool SidechainTxsCommitmentBuilder::add_fwt(const CTxForwardTransferOut& ccout,
+        LogPrint("sc", "%s():%d - scTxsCommitment building failed when adding FT for sidechain scId[%s] (CCTPlib error).\n",
+            __func__, __LINE__, ccout.GetScId().ToString());
+    }
+
```
We saw in the past some performance issues related to the printing of debug logs when dealing with the commitment tree. This is mainly because additions to the tree can happen quite frequently.

What do you think about printing the line only in case of error and doing nothing if the operation succeeds?

a-petrini replied:
Agree. Removed from both `SidechainTxsCommitmentBuilder` and `SidechainTxsCommitmentGuard`

a-petrini committed:
Addressing review comments

- Improved removal of empty elements from CommitmentGuard maps
- Improved encapsulation of CBlock::BuildScTxsCommitment()
- Improved log generation of SidechainTxsCommitmentBuilder and SidechainTxsCommitmentGuard: removed log when add is successful
- Added correct rewind of SidechainTxsCommitmentGuard after failure in CreateNewBlock()
- Change in fork height check
- Minor changes in comments

JackPiri commented on src/rpc/mining.cpp:
```diff
@@ -731,7 +731,10 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
         if (certSupported) {
             CCoinsViewCache view(pcoinsTip);
-            pblock->hashScTxsCommitment = pblock->BuildScTxsCommitment(view);
+            pblock->hashScTxsCommitment.SetNull();
```
Is it ok no call to `BuildScTxsCommitmentGuard` is performed here, compared for example to analogous code in `getblockmerkleroots`?

a-petrini replied:
Yes, it is safe, as we are building a commitment out a set of tx/cert that we know are safe, as they have been previously chosen by `CreateNewBlockWithKey()`, where a similar - actually more in-depth - check has been performed.

JackPiri commented on src/miner.cpp:
```diff
@@ -707,6 +727,48 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn,  unsigned int nBlo
+                }
+                if (!scCommitmentBuilderResult)
+                {
+                    LogPrint("sc", "%s():%d - Skipping [%s] because we cannot add that to the sc commitment tree\n",
```
I'd say also `scCommGuard` has to rebuilt (or its counters be updated) in case there is a failure on `scCommBuilder`.

a-petrini replied:
Yes, you are right! I modified this section to keep them aligned.

JackPiri commented on src/sc/sidechainTxsCommitmentGuard.cpp:
```diff
@@ -0,0 +1,280 @@
+    // as adding may fail for reasons different than having a full subtree.
+    if (counterRef.bwtr < cbs.BWTR_LIMIT) {
+        ++counterRef.bwtr;
+        LogPrint("sc", "%s():%d - scTxsCommitment guard: successfully added BWTR for sidechain scId[%s].\n",
```
(very minor importance) Maybe this info can be added  to the log: `- current size: %d.\n",`; at least to get aligned with `add_fwt`. Similar comment applies for `add_csw` and `add_cert`.

a-petrini replied:
Thanks, but as suggested by @ptagl  we removed all the logs concerning successful adds to the maps

JackPiri commented on src/gtest/test_libzendoo.cpp:
```diff
@@ -2491,3 +2501,578 @@ TEST(CctpLibrary, TestCustomFieldsValidationFix)
+    // Guard object must not contain anything about sidechainId2
+    ASSERT_EQ(guardObj.getCBS().cbsaMap.size(), 1);
+    ASSERT_EQ(guardObj.getCBS().cbsaMap.count(sidechainId),  1);
+    ASSERT_EQ(guardObj.getCBS().cbsaMap.count(sidechainId2), 0);
```
I'd insert just this additional check to ensure internal counters have been rewound correctly:

```
auto cbsRef  = guardObj.getCBS();
ASSERT_EQ(cbsRef.cbsaMap[sidechainId].ft, 10);
```

a-petrini replied:
Check added.

JackPiri commented on src/sc/sidechainTxsCommitmentGuard.cpp:
```diff
@@ -0,0 +1,280 @@
+        // Remove from the maps all the sidechains without entities
+        std::vector<uint256> aliveToBeRemoved;
+        for (auto [scid, cbsa]: cbs.cbsaMap) {
+            if ((cbsa.bwt == 0) && (cbsa.bwtr == 0) && (cbsa.cert == 0) && (cbsa.ft == 0))
```
(very minor importance) Just a matter a taste, would you consider this:

```
for (auto it = cbs.cbsaMap.cbegin(); it != cbs.cbsaMap.cend(); )
{
    if ((it->second.bwt == 0) && (it->second.bwtr == 0) && (it->second.cert == 0) && (it->second.ft == 0))
    {
        cbs.cbsaMap.erase(it++);
    }
    else
        ++it;
}
```

in order to avoid creating a vector for keeping track of the items to be removed?

Feel free to ignore if you prefer current style.

a-petrini replied:
I reimplemented this section in a very similar way to avoid the allocation of the vectors.

JackPiri commented on qa/rpc-tests/sc_big_commitment_tree_getblockmerkleroot.py:
```diff
@@ -0,0 +1,66 @@
+        4. Check if the ScTxsCommitmentTree stops updating due to limited height
+        '''
+
+        # Reach SC v2 fork height
```
I guess this comment should be `# Reach SC fork height`.

a-petrini replied:
Yes, corrected.

JackPiri reviewed (commented):
The PR looks very good and clean to me; I'm going to approve but please resolve this comment https://github.com/HorizenOfficial/zen/pull/515/files#r1124332139 before.
Thanks

i-Alex commented on src/sc/sidechainTxsCommitmentGuard.h:
```diff
@@ -0,0 +1,90 @@
+        static const int BWTR_LIMIT = 4095;
+        static const int CERT_LIMIT = 4095;
+        static const int  CSW_LIMIT = 4095;
+        static const int  BWT_LIMIT = 4096;
```
I think we need to keep BWT_LIMIT = 4095 as well because of the way how it is handled by CCTP lib (limit errors are ignored there)
Beside that it seems the limit is considered as a total number of BWT per sidechain, but must be per certificate.
For example, cert A (sc1, bt_number = 3000), cert B(sc1, bt_number 2000) are allowed to be put inside the same block. It is not an issue for CommTree.

a-petrini replied:
Yes, you are right. I changed the logic so that we check the limit per certificate. Also, limit has been decreased as suggested.

JackPiri reviewed (approved):
PR approved, thanks.

i-Alex commented on src/sc/sidechainTxsCommitmentGuard.h:
```diff
@@ -0,0 +1,94 @@
+        std::map<uint256, CommitmentBuilderStatsAliveCounter>  cbsaMap;
+        std::map<uint256, CommitmentBuilderStatsCeasedCounter> cbscMap;
+        // The following values MUST be aligned with those specified in CCTPlib!
+        static const int   SC_LIMIT = 4096;
```
Just a note: I've carefully double-checked the value for this limit, 4096 seems safe.
But in general, can we get all these constants from `zendoo-mc-cryptolib`?

a-petrini replied:
Thanks for checking! Getting the limits directly from mc-crypto or cctplib was the original idea, but unfortunately they require a rework on both libraries: their symbols are defined in cctplib, but not exposed to mc-crypto.
We plan to do this in another PR, however, to simplify the alignment of these limits.

a-petrini committed:
Addressing review comments 2

- Bwt limit decreased to 4095
- Bwt limit per cert and not for sc

i-Alex reviewed (approved):
LGTM!
